dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ProgramRuleEngine.java

```diff
@@ -191,25 +214,36 @@ public RuleValidationResult getDescription( String condition, Program program )
         return ruleEngine.evaluate( condition );
     }
 
-    private RuleEngineContext getRuleEngineContext( Program program, String uid )
+    private RuleEngineContext getRuleEngineContext( Program program, List<ProgramRule> programRules )
     {
         List<ProgramRuleVariable> programRuleVariables = programRuleVariableService
             .getProgramRuleVariable( program );
-        List<ProgramRule> programRules = implementableRuleService
-            .getImplementableRules( program )
-            .stream()
-            .filter( rule -> Objects.isNull( rule.getProgramStage() ) ||
-                Objects.equals( rule.getProgramStage().getUid(), uid ) )
-            .collect( Collectors.toList() );
 
         Map<String, String> constantMap = constantService.getConstantMap().entrySet()
             .stream()
             .collect( Collectors.toMap( Map.Entry::getKey, v -> v.getValue().toString() ) );
 
-        Map<String, List<String>> supplementaryData = organisationUnitGroupService.getAllOrganisationUnitGroups()
-            .stream()
-            .collect( Collectors.toMap( BaseIdentifiableObject::getUid,
-                g -> g.getMembers().stream().map( OrganisationUnit::getUid ).collect( Collectors.toList() ) ) );
+        List<String> orgUnitGroups = new ArrayList<>();
+
+        for ( ProgramRule programRule : programRules )
```

Comment:
Looks great. I think line 228 -237 could be moved to a Utils class and be united tested too.